test/NuGet.Core.Tests/NuGet.Packaging.Test/LicensesTests/NuGetLicenseExpressionParserTests.cs

```diff
@@ -0,0 +1,211 @@
+            }
+            else
+            {
+                Assert.Equal(ex.Message, string.Format(CultureInfo.CurrentCulture, Strings.NuGetLicenseExpression_ExceptionIdentifierIsLicense, badIdentifier));
+            }
+        }
+
+        [Theory]
+        [InlineData("(GPL-1.0 WITH 389-exception) OR MIT", "GPL-1.0")]
+        [InlineData("MIT OR GPL-1.0 WITH 389-exception AND Apache-2.0", "GPL-1.0")]
+        [InlineData("MIT OR GPL-1.0 WITH 389-exception", "GPL-1.0")]
+        [InlineData("MIT OR (LPL-1.0 OR GPL-1.0 WITH 389-exception)", "GPL-1.0")]
+        [InlineData("MIT OR (GPL-1.0 WITH 389-exception OR LPL-1.0)", "GPL-1.0")]
+        [InlineData("(LGPL-2.1 AND BSD-2-Clause)", "LGPL-2.1")]
+        [InlineData("((( (LGPL-2.1) AND BSD-2-Clause)))", "LGPL-2.1")]
+        public void LicenseExpressionParser_ComplexExpressionWithDeprecatedIdentifiersThrows(string expression, string deprecatedValue)
+        {
+            var ex = Assert.Throws<NuGetLicenseExpressionParsingException>(() => NuGetLicenseExpression.Parse(expression));
+            Assert.Equal(string.Format(CultureInfo.CurrentCulture, Strings.NuGetLicenseExpression_DeprecatedIdentifier, deprecatedValue), ex.Message);
+        }
+
+        [Theory]
+        [InlineData("MIT AND ")]
+        [InlineData("(MIT AND BSD-2-Clause")]
+        [InlineData("MIT (AND) LPL-1.0")]
+        [InlineData("((MIT) (AND) (LPL-1.0))")]
+        [InlineData("MIT (AND LPL-1.0)")]
+        [InlineData("MIT () OR LPL-1.0")]
+        [InlineData("MIT OR GPL-1.0 (WITH 389-exception OR LPL-1.0)")]
+        public void LicenseExpressionParser_StrictParseThrowsForInvalidExpressions(string expression)
+        {
+            Assert.Throws<NuGetLicenseExpressionParsingException>(() => NuGetLicenseExpression.Parse(expression));
+        }
+
+        [Theory]
+        [InlineData("MIT and BSD-2-Clause")]
+        [InlineData("MIT or BSD-2-Clause")]
+        [InlineData("MIT with Classpath-exception-2.0")]
+        public void LicenseExpressionParser_ThrowsForExpressionsWithBadCasing(string expression)
+        {
+            Assert.Throws<NuGetLicenseExpressionParsingException>(() => NuGetLicenseExpression.Parse(expression));
+        }
+
+        [Theory]
+        [InlineData("MIT WITH classpath-exception-2.0", "classpath-exception-2.0")]
+        public void LicenseExpressionParser_ThrowsForInvalidExceptionDueToBadCasing(string expression, string exception)
+        {
+            var ex = Assert.Throws<NuGetLicenseExpressionParsingException>(() => NuGetLicenseExpression.Parse(expression));
+            Assert.Equal(string.Format(CultureInfo.CurrentCulture, Strings.NuGetLicenseExpression_InvalidExceptionIdentifier, exception), ex.Message);
+        }
+
+        [Theory]
+        [InlineData("MIt WITH Classpath-exception-2.0", false, true)]
+        [InlineData("MIt AND BSD-2-Clause", false, true)]
+        [InlineData("MIt OR BSD-2-Clause", false, true)]
```

Comment:
test `+` case too

Reply:
Seems like you need a `true, false` and a `false, false` case too

Reply:
Also, seems like we need some tests for non-standard exceptions. Perhaps I missed them but I don't see any specific tests for that.

Reply:
The SPDX specification does not declare non-standard exception afaik.